src/meili/match_route.cc

```diff
@@ -122,39 +122,35 @@ bool MergeRoute(const State& source,
  * @param match_results  The matched points representing the matched input trace points
  * @param first_idx      The index of the first match in the range
  * @param last_idx       The index of the second match in the range
- * @param segments       The segments over the range
+ * @param segments_begin The current first segment for this range
+ * @param segments_end   The end of segments iterator
  * @param new_segments   The new vector of segments over the range (should be the same or more
  * segments). This is a reference parameter to avoid constant allocation
  */
 void cut_segments(const std::vector<MatchResult>& match_results,
                   int first_idx,
                   int last_idx,
-                  const std::vector<EdgeSegment>& segments,
+                  std::vector<EdgeSegment>::iterator& first_segment,
+                  const std::vector<EdgeSegment>::iterator& segments_end,
                   std::vector<EdgeSegment>& new_segments) {
-  auto first_segment = segments.begin();
   int prev_idx = first_idx;
   for (int curr_idx = first_idx + 1; curr_idx <= last_idx; ++curr_idx) {
     const MatchResult& curr_match = match_results[curr_idx];
     const MatchResult& prev_match = match_results[prev_idx];
 
-    // skip if we do not need to cut
-    if (!curr_match.is_break_point && curr_idx != last_idx) {
```

Comment:
so before this was just looping through non-interpolated points and only copied the segments when we were at the last segment. this would skip updating segments whose first or last matched point was interpolated. now there'll be a whole lot more allocations to the route vector below, only to keep the indices of the matched points correctly on the segments. in #3645 I tried to get around this by only adding segments to the route once we have a segment transition which was much less work for the algo. maybe I'll revisit that. what do you think?